src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java

```diff
@@ -287,55 +297,73 @@ public Config getConfig() {
-    * This pattern will then be converted to: <br/>
+    * This template will then be converted to: <br/>
     * <code>SELECT ?var1 ?var2 {?s ?var1 ?o . ?o &lt;http://exa.com&gt; ?var2}</code><br/>
     * and will request query solutions from the given sparql endpoint (e.g DBpedia).<br/>
-    * The solutions will then be instantiated into the query pattern.
+    * The solutions will then be instantiated into the template.
     * The result may look like the following:<br/>
     * <code>SELECT * WHERE {?s &lt;http://prop/1&gt; ?o . ?o &lt;http://exa.com&gt; "123"}</code><br/>
     * <code>SELECT * WHERE {?s &lt;http://prop/1&gt; ?o . ?o &lt;http://exa.com&gt; "12"}</code><br/>
     * <code>SELECT * WHERE {?s &lt;http://prop/2&gt; ?o . ?o &lt;http://exa.com&gt; "1234"}</code><br/>
     */
-    private static List<String> instantiatePatternQueries(QuerySource querySource, Config.Pattern config) throws IOException {
-        final var patternQueries = new FileCachingQueryList(querySource);
-        final Pattern pattern = Pattern.compile("%%var\\d+%%");
+    private static List<String> instantiateTemplateQueries(QuerySource querySource, Config.Template config) throws IOException {
+        // charset for generating random variable names
+        final String charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+        final Random random = new Random();
+
+        final var templateQueries = new FileCachingQueryList(querySource);
+        final Pattern template = Pattern.compile("%%var[0-9]+%%");
         final var instances = new ArrayList<String>();
-        for (int i = 0; i < patternQueries.size(); i++) {
-            // replace all variables in the query pattern with SPARQL variables
+        for (int i = 0; i < templateQueries.size(); i++) {
+            // replace all variables in the query template with SPARQL variables
             // and store the variable names
-            var patternQueryString = patternQueries.getQuery(i);
-            final Matcher matcher = pattern.matcher(patternQueryString);
-            final var variables = new ArrayList<String>();
+            var templateQueryString = templateQueries.getQuery(i);
+            final Matcher matcher = template.matcher(templateQueryString);
+            final var variables = new LinkedHashMap<String, String>(); // a set, that preserves insertion order
             while (matcher.find()) {
                 final var match = matcher.group();
-                final var variable = "?" + match.replaceAll("%%", "");
-                variables.add(variable);
-                patternQueryString = patternQueryString.replaceAll(match, variable);
+                if (variables.containsKey(match)) continue;
+                String variableName = match.replaceAll("%%", "");
+                while (templateQueryString.contains("?" + variableName)) { // generate random variable name with 20 characters until it is unique
```

Comment:
```suggestion
                while (templateQueryString.contains("?" + variableName) || templateQueryString.contains("$" + variableName)) { // generate random variable name with 20 characters until it is unique
```

$x and ?x are equivalent ways to write a variable. Don't ask me why they did this.